stob/ad.c

```diff
@@ -1171,11 +1174,12 @@ static void stob_ad_write_credit(const struct m0_stob_domain *dom,
 	 * of 'emap paste' (that is frags + 1) to verify this idea.
 	 */
 	m0_be_emap_credit(&adom->sad_adata, M0_BEO_PASTE, frags + 1, accum);
-
+#if 0
```

Comment:
It is not a good idea to disable this piece of code without any comments/explanation.

Please add detailed comments why it is disabled.

Reply:
Based on some older commit seems this piece of code was commentend out earlier as well and got uncommented 
as part of "4b8e109a299d8cb98ae3ba6b7ccb007ee7daa4b8" this commit id.
```

commit 4b8e109a299d8cb98ae3ba6b7ccb007ee7daa4b8
Author: Dmytro Podgornyi <dmytro.podgornyi@seagate.com>
Date:   Thu Oct 16 14:29:14 2014 +0000

    stob/ad: ad_overwrite field moved to m0_stob_ad_domain

    Make flag overwrite be domain-wide instead of stob-wide. This simlifies
    credit calculation for write operation, because doesn't require stob
    structure.
```

where it guarded under the condition (adom->sad_overwrite) based on assumption that 
(adom->sad_overwrite) is always false.(As it was initialize to false and never get set to True) 

```
    Change-Id: Ia575f38c62cea6ccafc2729fb96f69b158070363
-#if 0 /* Depends on as_overwrite flag which is always false now. */
-       /* for each emap_paste() seg_free() could be called 3 times */
-       if (ballroom->ab_ops->bo_free_credit != NULL)
+       if (adom->sad_overwrite && ballroom->ab_ops->bo_free_credit != NULL) {
+               /* for each emap_paste() seg_free() could be called 3 times */
                ballroom->ab_ops->bo_free_credit(ballroom, 3, accum);
-#endif
+       }
        m0_stob_io_credit(m0_stob_dom_get(adom->sad_bstore), io, accum);

```		
Hence we are disabling/commenting it out again as it was earlier, since we are setting this flag 
to true by default now for this fix. Will add comment and TODO into code as well.

Reply:
Based on some older commit seems this piece of code was commented out earlier as well and got uncommented 
as part of "4b8e109a299d8cb98ae3ba6b7ccb007ee7daa4b8" this commit id.

```
commit 4b8e109a299d8cb98ae3ba6b7ccb007ee7daa4b8
Author: Dmytro Podgornyi <dmytro.podgornyi@seagate.com>
Date:   Thu Oct 16 14:29:14 2014 +0000

    stob/ad: ad_overwrite field moved to m0_stob_ad_domain

    Make flag overwrite be domain-wide instead of stob-wide. This simlifies
    credit calculation for write operation, because doesn't require stob
    structure.
```
where it guarded under the condition (adom->sad_overwrite) based on assumption that  (adom->sad_overwrite) is always false.(As it was initialize to false and never get set to True) 

```
    Change-Id: Ia575f38c62cea6ccafc2729fb96f69b158070363
-#if 0 /* Depends on as_overwrite flag which is always false now. */
-       /* for each emap_paste() seg_free() could be called 3 times */
-       if (ballroom->ab_ops->bo_free_credit != NULL)
+       if (adom->sad_overwrite && ballroom->ab_ops->bo_free_credit != NULL) {
+               /* for each emap_paste() seg_free() could be called 3 times */
                ballroom->ab_ops->bo_free_credit(ballroom, 3, accum);
-#endif
+       }
        m0_stob_io_credit(m0_stob_dom_get(adom->sad_bstore), io, accum);
```
		
Hence we are disabling/commenting it out again as it was earlier, since we are setting this flag  to true by default now for this fix.

Reply:
So, is this `bo_free_credit()` call totally redundant? Can we safely call `stob_ad_seg_free` without reserving additional tx_credits?

Reply:
In existing code also(before this commit), we were never entering into if(adom->sad_overwite && ..)
As adom->sad_overwrite was always false, so it was anyway redundant earlier as well. Now we are setting this to true by default and that  may cause to enter into this if condition, if other condition is true.  So to maintain existing behaviour of stob_ad_credit_write(), disabling this piece of code with #if 0.
```

1188         if (adom->sad_overwrite && ballroom->ab_ops->bo_free_credit != NULL) {
1189                 /* for each emap_paste() seg_free() could be called 3 times */
1190                 ballroom->ab_ops->bo_free_credit(ballroom, 3 * frags, accum);
1191         }

```

>>>Can we safely call stob_ad_seg_free without reserving additional tx_credits?
That is I am not sure, but allowing to call stob_ad_seg_free() by setting sad_overwite True,
This issue got resolved, because stob_ad_seg_free() was never getting called since sad_overwite was always false.

Reply:
> Can we safely call stob_ad_seg_free without reserving additional tx_credits?
>> I am not sure

In this case, it would be good to add "We do not know whether bo_free_credit should be commented out or not." into the TODO comment.

Reply:
Updated comment in the code.